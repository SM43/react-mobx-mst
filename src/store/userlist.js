import { types, getEnv, flow } from "mobx-state-tree";
import { values } from "mobx";

export const User = types.model({
  id: types.integer,
  name: types.optional(types.string, ""),
  age: types.optional(types.integer, 0),
  class: types.optional(types.string, ""),
});

export const Store = types
  .model({
    users: types.array(User),
    filter: types.optional(types.array(types.string), []),
    sort: types.optional(types.string, ""),
    isLoading: true,
  })
  .views((self) => ({
    get count() {
      return self.users.length;
    },
    get userList() {
      return filterAndSortUsers(values(self.users), self.filter, self.sort);
    },
    get fetch() {
      return getEnv(self).fetch;
    },
  }))
  .actions((self) => ({
    afterCreate() {
      self.users.clear();
      loadUsers(self);
    },
    add(item) {
      self.users.push(item);
    },
    setSort(sort) {
      self.sort = sort;
    },
    setFilter(filter) {
      self.filter = filter;
    },
    setStatus(status) {
      self.isLoading = status;
    },
  }));

const loadUsers = flow(function* loadUsers(self) {
  try {
    const json = yield self.fetch("./users.json");
    json.forEach((userJson) => {
      self.add(userJson);
    });
    self.setStatus(false)
  } catch (err) {
    console.error("Failed to load users ", err);
  }
});

function filterAndSortUsers(users, filter, sort) {
  if (filter.length === 0 && sort === "") {
    return users;
  }
  if (filter.length === 0) {
    if (sort === "name") {
      return users.sort((a, b) =>
        a.name > b.name ? 1 : a.name === b.name ? 0 : -1
      );
    } else if (sort === "age") {
      return users.sort((a, b) =>
        a.age > b.age ? 1 : a.age === b.age ? 0 : -1
      );
    } else {
      return users;
    }
  }
  if (sort === "name") {
    return users
      .filter((u) => filter.includes(u.class))
      .sort((a, b) => (a.name > b.name ? 1 : a.name === b.name ? 0 : -1));
  } else if (sort === "age") {
    return users
      .filter((u) => filter.includes(u.class))
      .sort((a, b) => (a.age > b.age ? 1 : a.age === b.age ? 0 : -1));
  } else {
    return users.filter((u) => filter.includes(u.class));
  }
}
