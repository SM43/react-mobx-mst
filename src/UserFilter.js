import React from "react";
import { StoreContext } from "./store/store";

export const UserFilter = () => {
  const store = React.useContext(StoreContext);
  const filter = [];
  const setSort = (e) => {
    if (e.target.value === "Name") {
      store.setSort("name");
    } else if (e.target.value === "Age") {
      store.setSort("age");
    } else if (e.target.value === "No") {
      store.setSort("");
    }
  };
  const setClass = (e) => {
    if (
      e.target.value === "FY" ||
      e.target.value === "SY" ||
      e.target.value === "TY"
    ) {
      if (e.target.checked === true) {
        filter.push(e.target.value);
      } else {
        var index = filter.indexOf(e.target.value);
        while (index !== -1) {
          filter.splice(index, 1);
          index = filter.indexOf(e.target.value);
        }
      }
      store.setFilter(filter);
    }
  };
  return (
    <div>
      <h3>Filter: </h3>
      <div>
        <input
          type="checkbox"
          id="check1"
          defaultChecked={false}
          value="FY"
          onClick={(e) => {
            setClass(e);
          }}
        />{" "}
        FY
        <input
          type="checkbox"
          id="check2"
          defaultChecked={false}
          value="SY"
          onClick={(e) => {
            setClass(e);
          }}
        />{" "}
        SY
        <input
          type="checkbox"
          id="check2"
          defaultChecked={false}
          value="TY"
          onClick={(e) => {
            setClass(e);
          }}
        />{" "}
        TY
      </div>
      <div>
        <h3>Sort By:</h3>
        <input
          type="radio"
          id="r1"
          name="options"
          value="Name"
          onClick={(e) => {
            setSort(e);
          }}
        />{" "}
        Name
        <input
          type="radio"
          id="r2"
          name="options"
          value="Age"
          onClick={(e) => {
            setSort(e);
          }}
        />{" "}
        Age
        <input
          type="radio"
          id="r3"
          name="options"
          value="No"
          defaultChecked
          onClick={(e) => {
            setSort(e);
          }}
        />{" "}
        No Sort
      </div>
    </div>
  );
};
