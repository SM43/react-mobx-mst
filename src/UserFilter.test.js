import React from "react";
import renderer from "react-test-renderer";
import fs from "fs";
import { Store } from "./store/userlist";
import { StoreContext } from "./store/store";
import { UserFilter } from "./UserFilter.js";
import { mount } from "enzyme";

const userFetcher = () =>
  Promise.resolve(JSON.parse(fs.readFileSync("./src/store/users.json")));

const store = Store.create({}, { fetch: userFetcher });

const StoreProvider = ({ children }) => {
  return (
    <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
  );
};

describe("<UserFilter />", () => {
  it("should render correctly", () => {
    const component = renderer.create(<UserFilter />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});

it("should filter by class 'FY'", () => {
  const component = mount(
    <StoreProvider>
      <UserFilter />
    </StoreProvider>
  );
  component
    .find("#check1")
    .simulate("click", { target: { checked: true, value: "FY" } });
  const list = store.userList;
  expect(list.length).toBe(3);
  component
    .find("#check1")
    .simulate("click", { target: { checked: false, value: "FY" } });
  component.unmount();
});

it("should sort by name", () => {
  const component = mount(
    <StoreProvider>
      <UserFilter />
    </StoreProvider>
  );
  component.find("#r1").simulate("click");
  const list = store.userList;
  expect(list[0].name).toBe("chandler");
  component.find("#r1").simulate("click");
  component.unmount();
});
