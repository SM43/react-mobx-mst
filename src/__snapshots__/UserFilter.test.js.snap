// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserFilter /> should render correctly 1`] = `
<div>
  <h3>
    Filter: 
  </h3>
  <div>
    <input
      defaultChecked={false}
      id="check1"
      onClick={[Function]}
      type="checkbox"
      value="FY"
    />
     
    FY
    <input
      defaultChecked={false}
      id="check2"
      onClick={[Function]}
      type="checkbox"
      value="SY"
    />
     
    SY
    <input
      defaultChecked={false}
      id="check2"
      onClick={[Function]}
      type="checkbox"
      value="TY"
    />
     
    TY
  </div>
  <div>
    <h3>
      Sort By:
    </h3>
    <input
      id="r1"
      name="options"
      onClick={[Function]}
      type="radio"
      value="Name"
    />
     
    Name
    <input
      id="r2"
      name="options"
      onClick={[Function]}
      type="radio"
      value="Age"
    />
     
    Age
    <input
      defaultChecked={true}
      id="r3"
      name="options"
      onClick={[Function]}
      type="radio"
      value="No"
    />
     
    No Sort
  </div>
</div>
`;
